File: ios-im-ui/Sources/FlareIMUI/EmojiSticker/EmojiStickerViews.swift
import SwiftUI

private let bracketKey = try! NSRegularExpression(pattern: "^\\[([a-z][a-z0-9_]*)\\]$")
private let bareKey = try! NSRegularExpression(pattern: "^([a-z][a-z0-9_]*)$")
private let emojiToken = try! NSRegularExpression(pattern: "\\[([a-z][a-z0-9_]*)\\]")

private func firstGroup(_ re: NSRegularExpression, _ s: String) -> String? {
    let range = NSRange(s.startIndex..<s.endIndex, in: s)
    guard let m = re.firstMatch(in: s, range: range), m.numberOfRanges > 1,
          let r = Range(m.range(at: 1), in: s) else { return nil }
    return String(s[r])
}

private func resolvePackKey(_ raw: String) -> String? {
    let t = raw.trimmingCharacters(in: .whitespaces)
    return firstGroup(bracketKey, t) ?? firstGroup(bareKey, t)
}

private func flareResized(_ image: FlarePlatformImage, to side: CGFloat) -> FlarePlatformImage {
    #if canImport(UIKit)
    let renderer = UIGraphicsImageRenderer(size: CGSize(width: side, height: side))
    return renderer.image { _ in image.draw(in: CGRect(x: 0, y: 0, width: side, height: side)) }
    #else
    let out = NSImage(size: CGSize(width: side, height: side))
    out.lockFocus()
    image.draw(in: CGRect(x: 0, y: 0, width: side, height: side))
    out.unlockFocus()
    return out
    #endif
}

/// A static webp image from the bundled resource, with a fallback view.
public struct FlareBundleImage<Fallback: View>: View {
    private let url: URL?
    private let fallback: () -> Fallback
    public init(url: URL?, @ViewBuilder fallback: @escaping () -> Fallback) {
        self.url = url
        self.fallback = fallback
    }
    public var body: some View {
        if let url, let img = FlareEmojiStickerCatalog.shared.image(at: url) {
            Image(flarePlatformImage: img).resizable().scaledToFit()
        } else {
            fallback()
        }
    }
}

/// Emoji-pack message body (`[key]` / bare key / a raw unicode emoji).
public struct FlareEmojiPackMessage: View {
    private let emoji: String
    private let isSelf: Bool
    @Environment(\.colorScheme) private var scheme
    @Environment(\.locale) private var locale

    public init(emoji: String, isSelf: Bool = false) {
        self.emoji = emoji
        self.isSelf = isSelf
    }

    public var body: some View {
        let colors = FlareColors.of(scheme)
        if let key = resolvePackKey(emoji) {
            let label = FlareEmojiStickerCatalog.shared.emojiBracketLabel(key, locale: locale.identifier)
            FlareBundleImage(url: FlareEmojiStickerCatalog.shared.emojiImageURL(key)) {
                Text(label)
                    .font(.system(size: 20, weight: .medium))
                    .foregroundColor(colors.textSecondary)
                    .multilineTextAlignment(.center)
            }
            .frame(width: 120, height: 120)
        } else {
            Text(emoji).font(.system(size: 48))
        }
    }
}

/// Sticker message body — a bundled pack sticker (packageId + stickerId), else url.
public struct FlareStickerPackMessage: View {
    private let stickerId: String
    private let packageId: String?
    private let url: String?
    private let width: Int?
    private let height: Int?
    private let isSelf: Bool
    @Environment(\.colorScheme) private var scheme

    public init(stickerId: String, packageId: String? = nil, url: String? = nil,
                width: Int? = nil, height: Int? = nil, isSelf: Bool = false) {
        self.stickerId = stickerId
        self.packageId = packageId
        self.url = url
        self.width = width
        self.height = height
        self.isSelf = isSelf
    }

    public var body: some View {
        let colors = FlareColors.of(scheme)
        let maxSide: CGFloat = 120
        var w = CGFloat((width ?? 0) > 0 ? width! : 68)
        var h = CGFloat((height ?? 0) > 0 ? height! : 68)
        if w > maxSide || h > maxSide {
            let scale = maxSide / max(w, h)
            w *= scale; h *= scale
        }
        let bundleURL = stickerId.trimmingCharacters(in: .whitespaces).isEmpty
            ? nil
            : FlareEmojiStickerCatalog.shared.stickerImageURL(stickerId: stickerId, packageId: packageId)
        return FlareBundleImage(url: bundleURL) {
            RoundedRectangle(cornerRadius: FlareSizes.radiusLg)
                .fill(colors.bgHover)
                .overlay(RoundedRectangle(cornerRadius: FlareSizes.radiusLg).stroke(colors.borderPrimary))
                .overlay(Image(systemName: "face.smiling").foregroundColor(colors.textSecondary))
        }
        .frame(width: w, height: h)
    }
}

/// Renders plain text with inline `[key]` emoji images (unknown keys show their
/// localized label). Call only after excluding Markdown.
public struct FlarePlainTextEmojiRich: View {
    private let text: String
    private let font: Font
    private let color: Color?
    private let inlineSize: CGFloat
    @Environment(\.locale) private var locale

    public init(_ text: String, font: Font = .body, color: Color? = nil, inlineSize: CGFloat = 18) {
        self.text = text
        self.font = font
        self.color = color
        self.inlineSize = inlineSize
    }

    public var body: some View {
        composed().font(font).foregroundColor(color)
    }

    private func composed() -> Text {
        let catalog = FlareEmojiStickerCatalog.shared
        let ns = text as NSString
        let matches = emojiToken.matches(in: text, range: NSRange(location: 0, length: ns.length))
        if matches.isEmpty { return Text(text) }

        var result = Text("")
        var cursor = 0
        for m in matches {
            if m.range.location > cursor {
                result = result + Text(ns.substring(with: NSRange(location: cursor, length: m.range.location - cursor)))
            }
            let key = ns.substring(with: m.range(at: 1))
            if catalog.hasEmojiKey(key), let img = catalog.image(at: catalog.emojiImageURL(key)) {
                result = result + Text(Image(flarePlatformImage: flareResized(img, to: inlineSize)))
            } else {
                result = result + Text(catalog.emojiBracketLabel(key, locale: locale.identifier))
            }
            cursor = m.range.location + m.range.length
        }
        if cursor < ns.length {
            result = result + Text(ns.substring(from: cursor))
        }
        return result
    }
}

/// Composer emoji-pack + sticker picker. One tab for the emoji pack plus one per
/// sticker pack; taps emit `onInsertEmoji` / `onSendSticker`.
public struct FlareEmojiStickerPicker: View {
    private let onInsertEmoji: ((String) -> Void)?
    private let onSendSticker: ((_ packageId: String, _ stickerId: String) -> Void)?
    private let emojiLabel: String
    @State private var tab = 0
    @Environment(\.colorScheme) private var scheme

    public init(emojiLabel: String = "Emoji",
                onInsertEmoji: ((String) -> Void)? = nil,
                onSendSticker: ((_ packageId: String, _ stickerId: String) -> Void)? = nil) {
        self.emojiLabel = emojiLabel
        self.onInsertEmoji = onInsertEmoji
        self.onSendSticker = onSendSticker
    }

    public var body: some View {
        let colors = FlareColors.of(scheme)
        let catalog = FlareEmojiStickerCatalog.shared
        let packs = catalog.loadedStickerPacks()
        let current = min(max(tab, 0), packs.count)
        let columns = [GridItem(.adaptive(minimum: current == 0 ? 44 : 76), spacing: 8)]

        return VStack(spacing: 0) {
            ScrollView {
                LazyVGrid(columns: columns, spacing: 8) {
                    if current == 0 {
                        ForEach(catalog.loadedEmojiKeys(), id: \.self) { key in
                            FlareBundleImage(url: catalog.emojiImageURL(key)) { Color.clear }
                                .frame(width: 40, height: 40)
                                .onTapGesture { onInsertEmoji?(key) }
                        }
                    } else {
                        let pack = packs[current - 1]
                        ForEach(pack.stickerIds, id: \.self) { id in
                            FlareBundleImage(url: catalog.stickerImageURL(stickerId: id, packageId: pack.id)) { Color.clear }
                                .frame(width: 72, height: 72)
                                .onTapGesture { onSendSticker?(pack.id, id) }
                        }
                    }
                }
                .padding(10)
            }
            Divider()
            ScrollView(.horizontal, showsIndicators: false) {
                HStack(spacing: 6) {
                    let labels = [emojiLabel] + packs.map { $0.title }
                    ForEach(Array(labels.enumerated()), id: \.offset) { idx, label in
                        let selected = idx == current
                        Text(label)
                            .font(.system(size: FlareSizes.fontSizeSm, weight: selected ? .semibold : .regular))
                            .foregroundColor(selected ? colors.textPrimary : colors.textSecondary)
                            .padding(.horizontal, 12).padding(.vertical, 6)
                            .background(RoundedRectangle(cornerRadius: FlareSizes.radiusMd)
                                .fill(selected ? colors.bgHover : Color.clear))
                            .onTapGesture { tab = idx }
                    }
                }
                .padding(.horizontal, 8).padding(.vertical, 6)
            }
        }
        .frame(height: 300)
    }
}

File: android-im-ui/src/main/kotlin/com/flare/im/ui/EmojiStickerViews.kt
package com.flare.im.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.grid.GridCells
import androidx.compose.foundation.lazy.grid.LazyVerticalGrid
import androidx.compose.foundation.lazy.grid.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.SentimentSatisfiedAlt
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.platform.LocalConfiguration
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.Placeholder
import androidx.compose.ui.text.PlaceholderVerticalAlign
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.foundation.text.appendInlineContent
import androidx.compose.foundation.text.InlineTextContent
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import coil.ImageLoader
import coil.compose.AsyncImage
import coil.compose.SubcomposeAsyncImage

private val bracketKey = Regex("^\\[([a-z][a-z0-9_]*)]$")
private val bareKey = Regex("^([a-z][a-z0-9_]*)$")
private val emojiToken = Regex("\\[([a-z][a-z0-9_]*)]")

private fun resolvePackKey(raw: String): String? {
    val t = raw.trim()
    bracketKey.find(t)?.let { return it.groupValues[1] }
    bareKey.find(t)?.let { return it.groupValues[1] }
    return null
}

/** Remembers a Coil loader that animates webp, keyed on the application context. */
@Composable
fun rememberFlareEmojiStickerLoader(): ImageLoader {
    val ctx = LocalContext.current.applicationContext
    return remember(ctx) { flareEmojiStickerImageLoader(ctx) }
}

@Composable
private fun rememberCatalogLoaded(): Boolean {
    val ctx = LocalContext.current.applicationContext
    var loaded by remember { mutableStateOf(FlareEmojiStickerCatalog.isLoaded) }
    LaunchedEffect(Unit) {
        FlareEmojiStickerCatalog.ensureLoaded(ctx)
        loaded = true
    }
    return loaded
}

@Composable
private fun currentLocaleTag(): String =
    LocalConfiguration.current.locales.get(0)?.toLanguageTag() ?: "en"

/**
 * Emoji-pack message body (`[key]` / bare key / a raw unicode emoji). A known
 * pack key renders the animated webp; otherwise the localized `[label]` or a
 * large unicode glyph.
 */
@Composable
fun FlareEmojiPackMessage(emoji: String, isSelf: Boolean = false) {
    val colors = flareColors()
    val loaded = rememberCatalogLoaded()
    val locale = currentLocaleTag()
    val loader = rememberFlareEmojiStickerLoader()
    val packKey = resolvePackKey(emoji)

    if (packKey != null) {
        val label = if (loaded) FlareEmojiStickerCatalog.emojiBracketLabel(packKey, locale) else "[$packKey]"
        SubcomposeAsyncImage(
            model = FlareEmojiStickerCatalog.emojiAssetUri(packKey),
            imageLoader = loader,
            contentDescription = label,
            modifier = Modifier.size(120.dp),
            error = {
                Box(Modifier.size(120.dp), contentAlignment = Alignment.Center) {
                    Text(label, color = colors.textSecondary, fontSize = 20.sp, fontWeight = FontWeight.Medium)
                }
            },
        )
        return
    }
    Text(emoji, fontSize = 48.sp)
}

/**
 * Sticker message body — resolves a bundled pack sticker by `packageId` +
 * `stickerId`, falling back to a network url, then a placeholder.
 */
@Composable
fun FlareStickerPackMessage(
    stickerId: String,
    packageId: String? = null,
    url: String? = null,
    width: Int? = null,
    height: Int? = null,
    isSelf: Boolean = false,
) {
    val colors = flareColors()
    val loader = rememberFlareEmojiStickerLoader()
    val maxSide = 120
    var w = if ((width ?: 0) > 0) width!! else 68
    var h = if ((height ?: 0) > 0) height!! else 68
    if (w > maxSide || h > maxSide) {
        val scale = maxSide.toDouble() / (if (w > h) w else h)
        w = (w * scale).toInt()
        h = (h * scale).toInt()
    }

    val net = url?.trim().orEmpty()
    val model: Any = if (stickerId.trim().isNotEmpty()) {
        FlareEmojiStickerCatalog.stickerAssetUri(stickerId, packageId)
    } else {
        net
    }

    SubcomposeAsyncImage(
        model = model,
        imageLoader = loader,
        contentDescription = "贴纸",
        modifier = Modifier.size(w.dp, h.dp),
        error = {
            Box(
                Modifier
                    .size(w.dp, h.dp)
                    .clip(RoundedCornerShape(FlareSizes.radiusLg))
                    .background(colors.bgHover)
                    .border(1.dp, colors.borderPrimary, RoundedCornerShape(FlareSizes.radiusLg)),
                contentAlignment = Alignment.Center,
            ) {
                Icon(Icons.Outlined.SentimentSatisfiedAlt, contentDescription = null, tint = colors.textSecondary)
            }
        },
    )
}

// --- inline `[key]` emoji inside plain text -------------------------------

/**
 * Renders plain text with inline `[key]` emoji images (unknown keys show their
 * localized label). Call only after excluding Markdown.
 */
@Composable
fun FlarePlainTextEmojiRich(
    text: String,
    style: TextStyle = TextStyle.Default,
    inlineSize: Int = 20,
) {
    rememberCatalogLoaded()
    val locale = currentLocaleTag()
    val loader = rememberFlareEmojiStickerLoader()

    val matches = emojiToken.findAll(text).toList()
    if (matches.isEmpty()) {
        Text(text, style = style)
        return
    }

    val inline = mutableMapOf<String, InlineTextContent>()
    val annotated = buildAnnotatedString {
        var cursor = 0
        matches.forEachIndexed { i, m ->
            if (m.range.first > cursor) append(text.substring(cursor, m.range.first))
            val key = m.groupValues[1]
            if (FlareEmojiStickerCatalog.hasEmojiKey(key)) {
                val id = "emoji_$i"
                appendInlineContent(id, "[$key]")
                inline[id] = InlineTextContent(
                    Placeholder(
                        width = inlineSize.sp,
                        height = inlineSize.sp,
                        placeholderVerticalAlign = PlaceholderVerticalAlign.TextCenter,
                    ),
                ) {
                    AsyncImage(
                        model = FlareEmojiStickerCatalog.emojiAssetUri(key),
                        imageLoader = loader,
                        contentDescription = key,
                        modifier = Modifier.size(inlineSize.dp),
                    )
                }
            } else {
                append(FlareEmojiStickerCatalog.emojiBracketLabel(key, locale))
            }
            cursor = m.range.last + 1
        }
        if (cursor < text.length) append(text.substring(cursor))
    }
    Text(annotated, style = style, inlineContent = inline)
}

/**
 * Composer emoji-pack + sticker picker. One tab for the emoji pack plus one per
 * sticker pack; taps emit [onInsertEmoji] (`key` to insert as `[key]`) or
 * [onSendSticker] (`packageId`, `stickerId`).
 */
@Composable
fun FlareEmojiStickerPicker(
    onInsertEmoji: ((String) -> Unit)? = null,
    onSendSticker: ((packageId: String, stickerId: String) -> Unit)? = null,
    emojiLabel: String = "表情",
    modifier: Modifier = Modifier,
) {
    val colors = flareColors()
    val loaded = rememberCatalogLoaded()
    val loader = rememberFlareEmojiStickerLoader()
    var tab by remember { mutableStateOf(0) }

    if (!loaded) {
        Box(modifier.fillMaxWidth().height(300.dp), contentAlignment = Alignment.Center) {
            Text("…", color = colors.textTertiary)
        }
        return
    }

    val packs = FlareEmojiStickerCatalog.stickerPacks
    val current = tab.coerceIn(0, packs.size)

    Column(modifier.fillMaxWidth().height(300.dp)) {
        Box(Modifier.weight(1f)) {
            if (current == 0) {
                LazyVerticalGrid(
                    columns = GridCells.Adaptive(48.dp),
                    contentPadding = androidx.compose.foundation.layout.PaddingValues(10.dp),
                    horizontalArrangement = Arrangement.spacedBy(6.dp),
                    verticalArrangement = Arrangement.spacedBy(6.dp),
                ) {
                    items(FlareEmojiStickerCatalog.emojiKeys) { key ->
                        AsyncImage(
                            model = FlareEmojiStickerCatalog.emojiAssetUri(key),
                            imageLoader = loader,
                            contentDescription = key,
                            modifier = Modifier
                                .size(40.dp)
                                .clip(RoundedCornerShape(FlareSizes.radiusMd))
                                .clickable { onInsertEmoji?.invoke(key) }
                                .padding(4.dp),
                        )
                    }
                }
            } else {
                val pack = packs[current - 1]
                LazyVerticalGrid(
                    columns = GridCells.Adaptive(84.dp),
                    contentPadding = androidx.compose.foundation.layout.PaddingValues(10.dp),
                    horizontalArrangement = Arrangement.spacedBy(8.dp),
                    verticalArrangement = Arrangement.spacedBy(8.dp),
                ) {
                    items(pack.stickerIds) { id ->
                        AsyncImage(
                            model = FlareEmojiStickerCatalog.stickerAssetUri(id, pack.id),
                            imageLoader = loader,
                            contentDescription = id,
                            modifier = Modifier
                                .size(76.dp)
                                .clip(RoundedCornerShape(FlareSizes.radiusMd))
                                .clickable { onSendSticker?.invoke(pack.id, id) }
                                .padding(4.dp),
                        )
                    }
                }
            }
        }
        Row(
            Modifier
                .fillMaxWidth()
                .height(44.dp)
                .padding(horizontal = 8.dp, vertical = 6.dp),
            horizontalArrangement = Arrangement.spacedBy(6.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            val labels = listOf(emojiLabel) + packs.map { it.title }
            labels.forEachIndexed { i, label ->
                val selected = i == current
                Text(
                    label,
                    color = if (selected) colors.textPrimary else colors.textSecondary,
                    fontSize = FlareSizes.fontSizeSm.value.sp,
                    fontWeight = if (selected) FontWeight.SemiBold else FontWeight.Normal,
                    modifier = Modifier
                        .clip(RoundedCornerShape(FlareSizes.radiusMd))
                        .background(if (selected) colors.bgHover else androidx.compose.ui.graphics.Color.Transparent)
                        .clickable { tab = i }
                        .padding(horizontal = 12.dp, vertical = 6.dp),
                )
            }
        }
    }
}
